python_src/models/police_report.py

```diff
@@ -0,0 +1,251 @@
+class ConcreteCreatorC(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeC()
+
+
+class ConcreteCreatorD(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeD()
+
+
+class Report(ABC):
+    """
+    The Report interface declares the operations that all concrete reports
+    must implement.
+    """
+
+    @abstractmethod
+    def process(self, outputs: list, targets: list) -> Tuple[int, int, int]:
+        """
+        Return the amount of matched crashed elements(components/sides/parts),
+               the amount of matched non-crashed elements(components/sides/parts)
+               the amount of "matchable" elements for a specific report
+        """
+        pass
+
+    @staticmethod
+    def _match(category: list,
+                           t1: list, o1: list, t2: list, o2: list) -> Tuple[int, int]:
+        """
+        Count the amount of matched crashed elements(components/sides/parts) and
+              the amount of matched non-crashed elements(components/sides/parts)
+
+        Args:
+            t1 (list): List of expected CRASHED element from a police report
+            o1 (list): List of expected CRASHED element from a simulation
+            t2 (list): List of expected NON-CRASHED element from a police report
+            o2 (list): List of expected NON-CRASHED element from a simulation
+            category (list): Set of vehicle elements are used to compare
+
+        Returns:
+            A tuple contains the number of matched crashed
+            and non-crashed elements
+        """
+        crash_points = 0
+        non_crash_points = 0
+        for part in category:
+            # Count matching expected CRASHED component
+            crash_points += 1 if part in (set(t1) & set(o1)) else 0
+            # Count matching expected NON-CRASHED component
+            non_crash_points += 1 if part in (set(t2) & set(o2)) else 0
+
+        return crash_points, non_crash_points
+
+    @staticmethod
+    def _categorize_part(part: str, report_type: chr) -> chr:
+        """
+        Categorizes the type of vehicle part based on the type of police report.
+
+        Args:
+            part (str): The code name of vehicle part
+            report_type (chr): Type of report is used to categorize
+        """
+        if report_type == CAT_B:
+            return part[0]
+        elif report_type == CAT_C:
+            return part[1]
+        return None
+
+    @staticmethod
+    def _validate_output(outputs: list) -> True:
+        """
+        Verify the validity of given output from simulation
+        """
+        # Invalid given outputs
+        if len(outputs) is NO_CRASH:
+            raise Exception("Given scenario might not be crashed!")
+        # The vehicle's crash element should be on the car
+        parts = CAT_D_DATA
+        for item in [i["name"] for i in outputs]:
+            if item not in parts:
+                raise Exception(f'Part name {item} is not found!')
```

Comment:
I am not entirely sure about this. Do we trigger the exception at soon as there's one "unknown" element? or do we trigger the exception when there are only unknown crashed elements?
Also the message per see is not descriptive enough

Reply:
If there is one unknown element existed in the damage report of specific vehicle, it will trigger this exception. I changed the message to `The code {item} is not found in the part dictionary!`